File: Filtering in MVC 3/[C#]-Filtering in MVC 3/C#/Mvc3Filter/Filters/TraceActionFilterAttribute.cs
using System;
using System.Web.Mvc;
using System.Diagnostics;

namespace Mvc3Filter.Filters {
    public class TraceActionFilterAttribute : ActionFilterAttribute {
        public override void OnActionExecuting(ActionExecutingContext filterContext) {
            var txt = getResponseString(filterContext.ActionDescriptor, "OnActionExecuting");
            WriteResponse(txt, filterContext);
            base.OnActionExecuting(filterContext);
        }

        public override void OnActionExecuted(ActionExecutedContext filterContext) {
            var txt = getResponseString(filterContext.ActionDescriptor, "OnActionExecuted");
            WriteResponse(txt, filterContext);
            base.OnActionExecuted(filterContext);
        }

        public override void OnResultExecuting(ResultExecutingContext filterContext) {

            var txt = getResponseString(filterContext, "OnResultExecuting");
            WriteResponse(txt, filterContext);
        }

        public override void OnResultExecuted(ResultExecutedContext filterContext) {
            // Don't show filter multiple times when using Html.RenderAction or Html.Action.
            if (filterContext.IsChildAction == true)
                return;
            var txt = getResponseString(filterContext, "OnResultExecuted");
            WriteResponse(txt, filterContext);
        }

        string getResponseString(ControllerContext filterContext, string onAct) {
            var txt = String.Format(
                   "<hr /><h5>TraceActionFilterAttribute {0} : Action '{1} :: {2}'</h5>",
                   onAct,
                  filterContext.RouteData.Values["controller"],
                   filterContext.RouteData.Values["action"]
               );

            return txt;
        }
        string getResponseString(ActionDescriptor ad, string onAct) {
            var txt = String.Format(
                   "<hr /><h5>TraceActionFilterAttribute {0} : Action '{1} :: {2}'</h5>",
                   onAct,
                   ad.ControllerDescriptor.ControllerName,
                   ad.ActionName
               );

            return txt;
        }

        void WriteDbg( string txt) {
            Debug.WriteLine(txt);
        }

        void WriteResponse(string txt, ControllerContext filterContext) {
            var response = filterContext.HttpContext.Response;
            if (response.ContentType == "text/html")
                response.Write(txt);
            else
                Debug.WriteLine(txt);
        }
    }

}

File: Filtering in MVC 3/[VB]-Filtering in MVC 3/VB/Filters/TraceActionFilterAttribute.vb
Imports System.Web.Mvc

Namespace Mvc3Filter.Filters
	Public Class TraceActionFilterAttribute
		Inherits ActionFilterAttribute

		Public Overrides Sub OnActionExecuting(ByVal filterContext As ActionExecutingContext)
			Dim sFilterMethod As String = filterContext.GetType().FullName
			Dim txt = getResponseString(filterContext.ActionDescriptor, sFilterMethod)
			DebugInfo(filterContext, sFilterMethod)
			WriteResponse(txt, filterContext)
			MyBase.OnActionExecuting(filterContext)

			If filterContext.RouteData.Values.ContainsValue("Cancel") Then
				filterContext.Result = New RedirectResult("~/Home/Index")
				Trace.WriteLine(" Redirecting from Simple filter to /Home/Index")
			End If

		End Sub

		Private Sub DebugInfo(ByVal context As ControllerContext, ByVal sFilterMethod As String)
#If DbgTraceActionFilterAttribute Then
			DebugFilterHelper.DebugFilter(context, sFilterMethod, Me.ToString())
#End If
		End Sub

		Public Overrides Sub OnActionExecuted(ByVal filterContext As ActionExecutedContext)

			Dim sFilterMethod As String = filterContext.GetType().FullName
			DebugInfo(filterContext, sFilterMethod)
			Dim txt = getResponseString(filterContext.ActionDescriptor, sFilterMethod)

			WriteResponse(txt, filterContext)
			MyBase.OnActionExecuted(filterContext)
		End Sub

		Public Overrides Sub OnResultExecuting(ByVal filterContext As ResultExecutingContext)
			Dim sFilterMethod As String = filterContext.GetType().FullName
			DebugInfo(filterContext, sFilterMethod)
			Dim txt = getResponseString(filterContext, sFilterMethod)
			WriteResponse(txt, filterContext)
		End Sub

		Public Overrides Sub OnResultExecuted(ByVal filterContext As ResultExecutedContext)
			' Don't show filter multiple times when using Html.RenderAction or Html.Action.
			If filterContext.IsChildAction = True Then
				Return
			End If
			Dim sFilterMethod As String = filterContext.GetType().FullName
			DebugInfo(filterContext, sFilterMethod)

			Dim txt = getResponseString(filterContext, sFilterMethod)
			WriteResponse(txt, filterContext)
		End Sub

		Private Function getResponseString(ByVal filterContext As ControllerContext, ByVal onAct As String) As String
			Dim txt = String.Format("<hr /><h5>TraceActionFilterAttribute {0} : Action '{1} :: {2}'</h5>", onAct, filterContext.RouteData.Values(DebugFilterHelper.sController), filterContext.RouteData.Values(DebugFilterHelper.sAction))

			Return txt
		End Function
		Private Function getResponseString(ByVal ad As ActionDescriptor, ByVal onAct As String) As String
			Dim txt = String.Format("<hr /><h5>TraceActionFilterAttribute {0} : Action '{1} :: {2}'</h5>", onAct, ad.ControllerDescriptor.ControllerName, ad.ActionName)

			Return txt
		End Function

		Private Sub WriteDbg(ByVal txt As String)
			Trace.WriteLine(txt)
		End Sub

		Private Sub WriteResponse(ByVal txt As String, ByVal filterContext As ControllerContext)
			Dim response = filterContext.HttpContext.Response
			If response.ContentType = "text/html" Then
				response.Write(txt)
			Else
				Trace.WriteLine(txt)
			End If
		End Sub
	End Class

End Namespace